### PR TITLE
[In Progress] AI Integration: Adds Diagnostic Scope compatibility to emit Request Diagnostics

### LINKED ISSUE #3058: AI Integration: OTel CosmosDB Attributes to Collect
**Status:** [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md)
**Target SDK:** .Net SDK

## Operations To Cover
**Database / Container Operations :**  Cosmos.CreateDatabaseAsync / Cosmos.CreateContainerAsync / Cosmos.DeleteStreamAsync
**Point Operations :** Cosmos.CreateItemAsync / Cosmos.ReadItemAsync / Cosmos.UpsertItemAsync / Cosmos.ReplaceItemAsync / Cosmos.PatchItemAsync / Cosmos.DeleteItemAsync
**Stream Operations :**  Cosmos.CreateItemStreamAsync / Cosmos.UpsertItemStreamAsync / Cosmos.ReadItemStreamAsync / Cosmos.ReplaceItemStreamAsync / Cosmos.PatchItemStreamAsync / Cosmos.DeleteItemStreamAsync
**Batch Operations :** TBD
**Bulk Operation :** TBD
**Query Operations :** Cosmos.Typed FeedIterator ReadNextAsync (for each page)

#### AI Default Attributes:

Attributes |  Value |
-- | -- | 
Event time | 3/2/2022, 11:58:04.967 PM (Local time) |  
**Duration** | **278.0 ms** |  
**Name** | **Cosmos.DeleteStreamAsync** |  

Other default attributes are there e.g device details

##### Azure Specific Attributes
| Attribute  | Value | Comment |
| --------  | ------------------- | --------------------- |
| _kind_ | _client_       | _IGNORE, By Default, setting them as part of diagnostic scope_ | 
| _az.namespace_| _Microsoft.DocumentDB_  | _IGNORE,  By Default, setting them as part of diagnostic scope_  | 

##### Common Database Attributes
| Attribute  | Value | Comment |
| --------  | ------------------- | --------------------- |
|db.system| cosmosdb | Open Telemetry Convention To identify type of Db  ref. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#notes-and-well-known-identifiers-for-dbsystem|
| db.name | < Database Name >  |   | 
| db.operation| ReadItemAsync, DeleteItemStreamAsync etc  | Database Operation ~Type~ Name  | 
| net.peer.name |  e.g. sourabhjaintemp  |  Account Name + Cloud | 

##### Cosmos DB Specific
Account Level Information:
| Attribute  | Value | Comment |
| --------  | ------------------- | --------------------- |
| db.cosmosdb.client_id| Unique Client Id| Combination of client id and machine id can tell us, if customer is following best practices to create singleton client  |
| db.cosmosdb.machine_id| Unique Machine Id|  |
| user_agent.original | < User Agent With SDK version>  | Useful to identify the SDK version  | 
| db.cosmosdb.connection_mode | Direct/Gateway |   go through | 

Request Level Information:

| Attribute  | Value | Comment |
| --------  | ------------------- | --------------------- |
| db.cosmosdb.container| Container |    |
| db.cosmosdb.request_content_length_bytes | Size of request payload |    |
| db.cosmosdb.response_content_length_bytes | Size of response Payload |   | 
| db.cosmosdb.status_code  | 201/200/204   |  Cosmos Db Http Status Code, it tells if particular cosmosdb call/request is passed/failed with which HttpStatusCode  | 
| db.cosmosdb.sub_status_code  | 1000/1002   |  Cosmos Db SubStatus Code | 
| db.cosmosdb.request_charge  | < double type number > | RU consumed for that operation  | 
| db.cosmosdb.regions_contacted| Region Cosmos Db|  |
| db.cosmosdb.retry_count| Number of retries|  |
| db.cosmosdb.operation_type| Query/Read/Create |  |
| db.cosmosdb.item_count | < int number> |  Number of items returned by the operation, only Feed Operation |
| db.cosmosdb.request_diagnostics | < JSON String> | ~Open Question: What if this string is out of the limit size? will appinsight will break it and divide into different attributes?~ Generating it as event |
| db.cosmosdb.activity_id | Guid |  Unique Id for the operation and can be helpful to debug particular operation in backend logs  |
| db.cosmosdb.correlated_activity_id| Guid | It will be populated only in case of query operation to allow correlating query pages retrieved for the same multi-page or cross-partition query. |
| db.cosmosdb.batch_operations|  string | Comma separated list of operation type and count, only for batch operations|

##### Open Telemetry Standard for any Exception

| Attribute  | Value | Comment |
| --------  | ------------------- | --------------------- |
| exception.type| `java.net.ConnectException; ``OSError `| ref. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md | 
| exception.message| `Division by zero; Can't convert 'int' object to str implicitly `| ref. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md | 
| exception.stacktrace| `Exception in thread "main" java.lang.RuntimeException: Test exception\n at com.example.GenerateTrace.methodB(GenerateTrace.java:13)\n at com.example.GenerateTrace.methodA(GenerateTrace.java:9)\n at com.example.GenerateTrace.main(GenerateTrace.java:5)` | ref. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md | 

Following Open Telemetry Conventions (_**Note:** Status of below conventions is Experimental_):
~~1. https://github.com/Azure/azure-sdk/blob/main/docs/tracing/distributed-tracing-conventions.yml~~
2. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md
3. https://opentelemetry.io/docs/reference/specification/common/attribute-naming/

### PR DESCRIPTION
**Pull request by sourabh1007**
## Description

In this PR, `DiagnosticScope` is being used to emit Request Diagnostics object. 

PR related to emit Request Diagnostics Using `DiagnosticSource` is https://github.com/Azure/azure-cosmos-dotnet-v3/pull/2993

## Performance Impact (Using Open Telemetry)
With OpenTelemetry Listener , performance impact is significant. 
1. Unlike `DiagnosticSource`, Object gets serialized (to string at source) and de-serialized (at listener) and Object Serialization happens for each and every request even before Sampling or filtering happens.
**Suggested Solution:** Lazy serialization should happen.
2. Total performance impact depends on the listener's implementation.

Baselined Avg RPS is : 11500 (master branch)

Without any Exporter and Sampling :  7658 (33% impact, Serialization is happening for all the requests)
With Just Exporter : ~7500 (~33% impact, Serialization is happening for all the requests)

Different Ways to reduce the impact:
### Sampling
   Using [TraceIdRatioBasedSampler](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Trace/TraceIdRatioBasedSampler.cs):  
<u>Attempt 1:</u>
`traceBuilder.SetSampler(new TraceIdRatioBasedSampler(<number between 0 and 1>));`
a) 1 out of 1000   - Avg PRS 8198 ( around 2k appinsight calls, Serialization is happening for all the requests)
b) 1 out of 10000 - Avg RPS 8175 (around 200 appinsight calls, Serialization is happening for all the requests)
Avg RPS is always around 8100, no matter what configuration is. It is around ~18% impact on performance.<br>
<u>Attempt 2:</u> With `Activity.Current.IsAllDataRequested` check around adding attribute
a) 1 out 10000 - Avg RPS is 10561 (Creates around ~200 entries in AppInsight, Serialization happens Same number of time)
b) 1 out of 1000 - Avg RPS is 10432 (Creates around ~2000 entries in AppInsight, Serialization happens Same number of time)
Performance impact is around ~10%

### Filtering at Source (Without Sampling, Collecting all events but sending diagnostics only for few)
<table style="width: 275px;" width="440">
<thead>
<tr>
<td style="width: 265.8px; text-align: center;" colspan="2"><strong>Without Listener</strong></td>
</tr>
<tr>
<td style="width: 165.663px;">% Request Diagnostics</td>
<td style="width: 94.5375px;">Avg RPS</td>
</tr>
</thead>
<tbody>
<tr>
<td style="width: 165.663px;">N/A</td>
<td style="width: 94.5375px;">11041</td>
</tr>
</tbody>
</table>

<table style="height: 54px; width: 269px;" width="440">
<tbody>
<tr style="height: 18px;">
<td style="height: 18px; width: 259.8px; text-align: center;" colspan="2"><b>Listener + Serialization at SDK </b></td>
</tr>
<tr style="height: 18px;">
<td style="height: 18px; width: 168.55px;">% Request Diagnostics</td>
<td style="height: 18px; width: 85.65px;">Avg RPS</td>
</tr>
<tr style="height: 18px;">
<td style="height: 18px; width: 168.55px;">100</td>
<td style="height: 18px; width: 85.65px;">8600</td>
</tr>
</tbody>
</table>

**Analysis:** You can notice RPS dropped steeply impact is more than 20% if we do send all the request with serialization at SDK.

<table style="width: 311px;" width="440">
<tbody>
<tr>
<td style="width: 301.8px;" colspan="2"><strong>Listener + Serialization it NOT at SDK</strong></td>
</tr>
<tr>
<td style="width: 170.688px;">% Request Diagnostics</td>
<td style="width: 125.512px;">Avg RPS</td>
</tr>
<tr>
<td style="width: 170.688px;">100</td>
<td style="width: 125.512px;">10500</td>
</tr>
</tbody>
</table>
**Analysis:** You can notice RPS dropped by 5% around because Serialization is not happening at all. Listener is just listening the data with acting on it.

<table style="width: 369px;" width="440">
<tbody>
<tr>
<td style="width: 359.8px; text-align: center;" colspan="2"><strong>Listener + Exporter + Serialization at SDK</strong></td>
</tr>
<tr>
<td style="width: 236.887px; text-align: center;">% Request Diagnostics</td>
<td style="width: 117.312px; text-align: center;">Avg RPS</td>
</tr>
<tr>
<td style="width: 236.887px; text-align: center;">50</td>
<td style="width: 117.312px; text-align: center;">8923</td>
</tr>
<tr>
<td style="width: 236.887px; text-align: center;">25</td>
<td style="width: 117.312px; text-align: center;">9454</td>
</tr>
<tr>
<td style="width: 236.887px; text-align: center;">12.5</td>
<td style="width: 117.312px; text-align: center;">10016</td>
</tr>
<tr>
<td style="width: 236.887px; text-align: center;">0</td>
<td style="width: 117.312px; text-align: center;">10735</td>
</tr>
</tbody>
</table>

**Analysis:** It is an attempt to reduce the impact due to serialization at SDK. (_Note: It has impact due to exporter also_). So we can see that performance  impact is direct proportional to the number of requests diagnostics you are serializing.

<table style="margin-left: auto; margin-right: auto;" width="440">
<tbody>
<tr>
<td style="text-align: center;" colspan="2" width="440"><strong>Listener + Exporter + Serialization it NOT at SDK</strong></td>
</tr>
<tr style="text-align: center;">
<td>% Request Diagnostics</td>
<td>Avg RPS</td>
</tr>
<tr style="text-align: center;">
<td>50</td>
<td>10028</td>
</tr>
<tr style="text-align: center;">
<td>12.5</td>
<td>10103</td>
</tr>
<tr style="text-align: center;">
<td>5</td>
<td>10212</td>
</tr>
</tbody>
</table>

**Analysis:** This time leave the serialization responsibility to Listener/exporter. We can notice the improvement in performance.

#### Conclusion on the basis of above numbers and further analysis is:
_Master Code_ Vs _DiagnosticScope W/o Listener_ = 1-2% of impact
_DiagnosticScope W/o Listener_ Vs _With open telemetry Listener_ = around 10% of impact (it can vary with listener)
**_So Overall Performance impact is 12-15% around._**

### Using Open Telemetry

```
using var openTelemetry = Sdk.CreateTracerProviderBuilder()
    .AddSource("Azure.*") // Collect all traces from Azure SDKs
    .AddAzureMonitorTraceExporter(options => options.ConnectionString = 
        "<Application Insights connection string>") // Export traces to Azure Monitor
    .Build(); // Start listening
```
Ref: https://devblogs.microsoft.com/azure-sdk/introducing-experimental-opentelemetry-support-in-the-azure-sdk-for-net/

Flexibility to use any exporter, ref https://opentelemetry.io/docs/instrumentation/net/exporters/

### Using Custom Diagnostic Listener
A listener like this https://github.com/Azure/azure-cosmos-dotnet-v3/blob/88d16de852cbede991967a7c8287dfac0156aad5/Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Diagnostics/ClientDiagnosticListener.cs

Needs to be initialize like this : 
`using var testListener = new ClientDiagnosticListener("Azure.Cosmos");`

## Application Insight View

Refer #3059 

## Type of change
- [] New feature (non-breaking change which adds functionality)

## Review Guidance
Since this is a big PR with lots of file changes So adding this section in order to provide guidance so that main functionality related code changes can be reviewed on priority.

#### Open to discuss:
1. Design Decisions : Right Now I decided to keep it in Trace as it already flows in the code. it is easy to collect relevant information. 
2. Attributes to collect : https://github.com/Azure/azure-cosmos-dotnet-v3/blob/1cdc98c31f861a11296c5a956f88a3cd9fc563d9/Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs#L149
Open to add/remove any attribute.

####  Main Code Changes
1.  `Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs` : It has the main code related to instrumentation. From here you can notice that Instrumentation is part of Trace.
2. https://github.com/Azure/azure-cosmos-dotnet-v3/blob/0e61bc42f3280f287e02b57d0e9bdcaeb59d3127/Microsoft.Azure.Cosmos/src/Tracing/Trace.cs#L157 : Instrumentation is getting initialized here with Traces.
3. `Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs` recording happening in this class.
4. For bulk NoOpTrace is there.

#### Performance Tests
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/0e61bc42f3280f287e02b57d0e9bdcaeb59d3127/Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs#L54 : It is where open telemetry is initialized

#### Unit Test Cases
1. `Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Telemetry/OpenTelemetryTests.cs`
2. `Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Telemetry/ClientDiagnosticListener.cs`

#### Things to Ignore:
1. You can find few Console statements.
2. few version change and new libraries import

**Note:** Perf test did not run for current code. Once we achieved what User experience We are targeting then I will run perf test for final perf testing.

Closes #3058 and #3059

### CODE REVIEW

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs [hunk 1]
```diff
@@ -45,7 +48,32 @@ public static async Task Main(string[] args)
...
+                TracerProvider openTelemetry = null;
+                if (config.EnableOpenTelemetry)
+                {
+                    AppContext.SetSwitch("Azure.Experimental.EnableActivitySource", true);
+                    TracerProviderBuilder traceBuilder = Sdk.CreateTracerProviderBuilder()
+                                            .AddSource("Azure.*"); // Collect all traces from Azure SDKs
+                    if (config.AppInsightConnectionString != null)
+                    {
+                        traceBuilder
+                            .AddAzureMonitorTraceExporter(
```
Comment (lmolkova) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:60]: it might be beneficial to turn off real reporting to appinsights for perf testing - i'd expect a lot of noise from ingestion service RPC calls and other exporter features, especially if it's used on a huge amount of data with sampling off (seems to be commented out below).

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs [hunk 2]
```diff
@@ -455,33 +461,70 @@ protected virtual void Dispose(bool disposing)
...
-                catch (NullReferenceException nullRefException) when (!(nullRefException is CosmosNullReferenceException))
-                {
-                    throw new CosmosNullReferenceException(
-                        nullRefException,
-                        trace);
-                }
+            {
+                using (DiagnosticScope scope = CosmosDbInstrumentation.ScopeFactory.CreateScope($"Cosmos.{operationName}"))
+                {
+                    try
```
Comment (lmolkova) [inline Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:473]: can you please save `Activity.Current != null && scope.IsEnabled && Activity.Current.IsAllDataRequested` into local variable and use it instead or re-calculating it everywhere? `Activity.Current` is `AsyncLocal` and even though accessing it is relatively cheap, it makes sense to avoid extensive use of it.

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs [hunk 3]
```diff
@@ -455,33 +461,70 @@ protected virtual void Dispose(bool disposing)
...
+                    {
+                        if (Activity.Current != null && scope.IsEnabled && Activity.Current.IsAllDataRequested)
+                        {
+                            int percentToAllow = 3;
+                            int totalItems = 2000000;
+                            int allowedItems = percentToAllow * totalItems / 100;
+                            int randomNumber = this.random.Next(1, totalItems);
+                            if (randomNumber <= allowedItems)
+                            {
+                                scope.AddAttribute("Request Diagnostics", new CosmosTraceDiagnostics(trace));
```
Comment (lmolkova) [inline Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:523]: As we discussed, [complex attributes types](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attribute) are not really supported by opentelemetry and it works by accident.
 
since you're already using DiagnosticSource 6, you can use Activity.AddEvent and pass serialized `CosmosTraceDiagnostics` as an event name. This would also help to make sure we're perf-testing the right path (but I don't expect any perf-improvements from it)

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/BenchmarkConfig.cs [hunk 4]
```diff
@@ -120,6 +120,15 @@ public class BenchmarkConfig
...
         public string ResultsContainer { get; set; } = "runsummary";
 
+        [Option(Required = false, HelpText = "Enable Open Telemetry")]
+        public bool EnableOpenTelemetry { get; set; }
+
+        [Option(Required = false, HelpText = "Open Telemetry AppInsights Connection String")]
+        public string AppInsightConnectionString { get; set; }
+
+        [Option(Required = false, HelpText = "Open Telemetry Sample Allowed Ratio")]
+        public double SampleAllowedRatio { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/BenchmarkConfig.cs:130]: How about SamplingRatio or SamplingCollectonRatio

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/BenchmarkConfig.cs [hunk 5]
```diff
@@ -120,6 +120,15 @@ public class BenchmarkConfig
         [Option(Required = false, HelpText = "Container to publish results to")]
         public string ResultsContainer { get; set; } = "runsummary";
 
+        [Option(Required = false, HelpText = "Enable Open Telemetry")]
+        public bool EnableOpenTelemetry { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/BenchmarkConfig.cs:124]: Telemetry as name is overloaded in our case in two scenarios. Do we want to name it more specific or 'OpenTelemetry' is to dis-ambiguate?

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/JsonHelper.cs [hunk 6]
```diff
@@ -15,7 +15,7 @@ internal static class JsonHelper
                     NullValueHandling = NullValueHandling.Ignore,
                     Formatting = Formatting.Indented,
                 });
-        private const int DefaultCapacity = 1024;
+        private const int DefaultCapacity = 2048;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/JsonHelper.cs:18]: Can you please expand on this change?

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs (hunk 1 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:60]: One alternative is to do one more extra run with-out exporter and use that for the regression analysis.
With exporter is more for as our data point.

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/BenchmarkConfig.cs (hunk 4 above)
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/BenchmarkConfig.cs:130]: Nullable to identify if its set or not?

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs [hunk 7]
```diff
@@ -45,7 +48,31 @@ public static async Task Main(string[] args)
...
                 Program program = new Program();
 
-                RunSummary runSummary = await program.ExecuteAsync(config);
+                TracerProvider openTelemetry = null;
+                if (config.EnableOpenTelemetry)
+                {
+                    AppContext.SetSwitch("Azure.Experimental.EnableActivitySource", true);
+                    TracerProviderBuilder traceBuilder = Sdk.CreateTracerProviderBuilder()
+                                            .AddSource("Azure.*"); // Collect all traces from Azure SDKs
+                   /* if (config.AppInsightConnectionString != null)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:57]: Are these still expected to be commented?

Also for config strings general practice is to use String.IsNullOrWhite....

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs [hunk 8]
```diff
@@ -45,7 +48,31 @@ public static async Task Main(string[] args)
 
                 Program program = new Program();
 
-                RunSummary runSummary = await program.ExecuteAsync(config);
+                TracerProvider openTelemetry = null;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:51]: Is it expected to be used outside below if scope? 

If not then it can be inlined with the assignment

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs [hunk 9]
```diff
@@ -45,7 +48,31 @@ public static async Task Main(string[] args)
...
+                    }*/
+                    
+                    /* if (config.SampleAllowedRatio > 0)
+                    {
+                        traceBuilder.SetSampler(new TraceIdRatioBasedSampler(config.SampleAllowedRatio));
+                    }*/
+
+                    openTelemetry = traceBuilder.Build();
+                }
+                using (openTelemetry)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:72]: Isn't this needs to go into the above if-conditional block?

File: Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+#nullable enable
+
+namespace Azure.Core
+{
+    using System;
+    /// <summary>
+    /// Helper for interacting with AppConfig settings and their related Environment variable settings.
+    /// </summary>
+    internal static class AppContextSwitchHelper
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs:13]: How about move it to Utils?

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 11]
```diff
@@ -120,6 +120,7 @@
 
 	<ItemGroup>
 		<PackageReference Include="System.Collections.Immutable" Version="1.7.0" />
+		<PackageReference Include="System.Diagnostics.DiagnosticSource" Version="6.0.0" />
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj:123]: Liudmila suggested to list this dependency as 4.6.0 (for some functions compatibility)

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs (hunk 7 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:57]: uncommented it...it was commented for testing purpose

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs (hunk 8 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:51]: it is expected ouside the scope...because `executeSysnc` should be called even when open telemetry is not initialized.

File: Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs (hunk 9 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs:72]: no...executesync should be called even when openTelemetry is not Enabled.

File: Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs (hunk 10 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs:13]: I just wanted to keep all the AzureCore related files at one place. Instead of scattering them in the code base.

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs (hunk 3 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs]: Downgraded to DiagnosticSource 4.6.0...

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 12]
```diff
@@ -174,11 +175,33 @@ internal class RequestInvokerHandler : RequestHandler
...
-                                    return dce.ToCosmosResponseMessage(request);
+                                    response = dce.ToCosmosResponseMessage(request);
+
+                                    trace.CosmosInstrumentation.RecordWithException(requestCharge: response?.Headers?.RequestCharge,
+                                                                       operationType: operationType.ToOperationTypeString(),
+                                                                       statusCode: response.StatusCode,
+                                                                       containerId: cosmosContainerCore?.Id,
+                                                                       databaseId: cosmosContainerCore?.Database?.Id,
+                                                                       exception: dce,
+                                                                       subStatusCode: response?.Headers?.SubStatusCode.ToSubStatusCodeString(),
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:186]: Why is this needed? The dce already has the substatus code. This should be logged as an int and not the enum string value. If a new substatus code is sent, then it will be logged as unknown when it could be an actual value.

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs [hunk 13]
```diff
@@ -463,25 +462,36 @@ protected virtual void Dispose(bool disposing)
             {
                 try
                 {
+                    trace.CosmosInstrumentation.Record(this.client?.Endpoint, this.UserAgent, this.ClientOptions.ConnectionMode);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:465]: Is this getting logged for every request?

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs [hunk 14]
```diff
@@ -463,25 +462,36 @@ protected virtual void Dispose(bool disposing)
...
+                catch (NullReferenceException nullRefException) when
+                    (!(nullRefException is CosmosNullReferenceException))
                 {
                     throw new CosmosNullReferenceException(
                         nullRefException,
                         trace);
                 }
+                finally
+                {
+                    trace.CosmosInstrumentation.Record(new CosmosTraceDiagnostics(trace));
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:490]: This is a circular dependency. Why does trace need to record the CosmosInstrumentation?

File: Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs (hunk 13 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs:465]: yes...it will get logged on each and every  request.

**Comment by sourabh1007** [comment]
Closing this PR as these changes has been merged in SDK through different small PRs
